src/main/scala/org/wikimedia/analytics/hdfstools/HdfsRsyncExec.scala

```diff
@@ -259,84 +280,142 @@ class HdfsRsyncExec(config: HdfsRsyncConfig) {
-            (!config.sizeOnly &&                           // (Use modificationTimes AND
-                approxDiffTimestamps(src, existingTarget)) //     modificationTimes are different)
+        config.ignoreTimes ||                                      // Force copy/update OR
+            src.getLen != existingTarget.getLen ||                 // Size is different OR
+            (!config.sizeOnly &&                                   // (Use modificationTimes AND
+                approxCompareTimestamps(src, existingTarget) == 0) // modificationTimes are different)
     }
 
     /**
-     * Function processing a single src file, with its associated target in dst, and
+     * Function processing a single src file, with its associated target in dst, and its
      * potential existingTarget for reference to the possibly already existing object in dst.
-     * The src file will be copied if new, updated if already existing and different,
-     * or skipped if already existing and not different (configuration dependent).
-     * Once target is up-to-date with src, permissions and modificationTimes are
-     * updated if configuration says so.
+     * The target file will be copied if new, updated if already existing and different,
+     * or omitted if already existing and not different (configuration dependent).
+     *
+     * The returned value will be applied metadata changes (modification-times, permissions
+     * and ownership). Some is returned when the file is copied or omitted because no
+     * change is needed, None when the file skipped  because of configuration settings.
      *
      * @param src the src FileStatus to serve as base
      * @param target the target for the possibly copied/updated files.
      * @param existingTarget the reference to the already existing file at target
      *                       path in dst, if any.
+     * @return Left(Path) of target in case of skip or dryrun (no FileStatus available),
+     *         Right(FileStatus) of the processed target otherwise
      */
-    private def processFile(
+    private def copyAsNeeded(
         src: FileStatus,
-        target: Option[Path],
-        existingTarget: Option[FileStatus]
-    ): Unit = {
-        // If no target is provided, just log
-        if (target.isEmpty) {
-            log.info(s"COPY_FILE [no-dst] - ${src.getPath}")
-        } else {
-            val isNew = existingTarget.isEmpty
-            val targetToUpdate = {
-                if (isNew) {
-                    copyNewFile(src, target.get)
-                } else if (areDifferent(src, existingTarget.get)) {
-                    updateFile(src, target.get)
+        target: Path,
+        existingTarget: Option[FileStatus],
+        isNew: Boolean
+    ): Option[Either[Path, FileStatus]] = {
+        // Skipping facility
+        def skip(
+            msgTag: String,
+            res: Option[Either[Path, FileStatus]] = None
```

Comment:
I think you were going to remove the res return value and make this just a dump logging helper/wrapper, right?  You only ever call this function wtih res=None now, and it would be more readable to just call this and return None explicitly.

e.g.

```
logSkip("existing")
None
```